File: backend/app/schemas/loan.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from ..models import ApplicationStatus


class LoanProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: Optional[str]
    min_amount: Decimal
    max_amount: Decimal
    interest_rate: Decimal
    repayment_period_days: int
    # CLAUDE.md §19 upfront term transparency: shown to the customer before
    # they apply, not just derived after approval.
    interest_model: str
    installment_count: int


class LoanApplicationCreateRequest(BaseModel):
    loan_product_id: int
    amount_requested: Decimal = Field(gt=0)


class LoanApplicationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    loan_product_id: int
    amount_requested: Decimal
    status: ApplicationStatus
    review_notes: Optional[str]
    reviewed_at: Optional[datetime]
    created_at: datetime


class RepaymentInstallmentResponse(BaseModel):
    id: int
    installment_number: int
    due_date: date
    amount_due: Decimal
    amount_paid: Decimal
    # CLAUDE.md §19 borrower transparency: principal vs interest, not just a
    # lump "amount due". principal_component + interest_component == amount_due.
    principal_component: Decimal
    interest_component: Decimal
    is_paid: bool


class CustomerLoanResponse(BaseModel):
    id: int
    loan_product_name: str
    principal: Decimal
    interest_rate: Decimal
    total_repayable: Decimal
    # CLAUDE.md §23: 3-way breakdown — principal, base interest
    # (total_repayable - principal), and penalties, never one opaque number.
    penalties_accrued: Decimal
    outstanding_balance: Decimal
    status: str
    disbursed_at: Optional[datetime]
    created_at: datetime
    schedule: list[RepaymentInstallmentResponse]


class RepaymentRequest(BaseModel):
    amount: Decimal = Field(gt=0)


class RepaymentResponse(BaseModel):
    loan_id: int
    amount_paid_now: Decimal
    outstanding_balance: Decimal
    status: str


class CustomerCreditSummaryResponse(BaseModel):
    # CLAUDE.md §9/§1: no credit scoring exists yet — `loan_limit` is not a
    # score, just the largest amount any active product in this company
    # offers. `available_credit` nets out what the customer currently owes
    # on active loans. A real scoring model (advisory only, per rule #8) can
    # replace this heuristic later without changing the response shape.
    loan_limit: Decimal
    available_credit: Decimal
    standing: str
    loans: list[CustomerLoanResponse]

File: frontend/src/schemas/loan.ts
// Mirrors backend app/schemas/loan.py
export type InterestModel = "flat" | "reducing_balance" | "daily_accrual";

export interface LoanProductResponse {
  id: number;
  name: string;
  description: string | null;
  min_amount: string;
  max_amount: string;
  interest_rate: string;
  repayment_period_days: number;
  interest_model: InterestModel;
  installment_count: number;
}

export interface LoanApplicationResponse {
  id: number;
  loan_product_id: number;
  amount_requested: string;
  status: "pending" | "approved" | "rejected";
  review_notes: string | null;
  reviewed_at: string | null;
  created_at: string;
}

export type LoanStatus = "approved" | "active" | "overdue" | "defaulted" | "repaid";

export interface RepaymentInstallmentResponse {
  id: number;
  installment_number: number;
  due_date: string;
  amount_due: string;
  amount_paid: string;
  // CLAUDE.md §19: principal vs interest breakdown — principal_component +
  // interest_component === amount_due.
  principal_component: string;
  interest_component: string;
  is_paid: boolean;
}

export interface CustomerLoanResponse {
  id: number;
  loan_product_name: string;
  principal: string;
  interest_rate: string;
  total_repayable: string;
  // CLAUDE.md §23: 3-way breakdown — principal, base interest
  // (total_repayable - principal), and penalties, never one opaque number.
  penalties_accrued: string;
  outstanding_balance: string;
  status: LoanStatus;
  disbursed_at: string | null;
  created_at: string;
  schedule: RepaymentInstallmentResponse[];
}

export interface CustomerCreditSummaryResponse {
  loan_limit: string;
  available_credit: string;
  standing: "good_standing" | "attention_needed";
  loans: CustomerLoanResponse[];
}

export interface RepaymentResponse {
  loan_id: number;
  amount_paid_now: string;
  outstanding_balance: string;
  status: LoanStatus;
}

export interface TransactionResponse {
  id: number;
  loan_id: number;
  type: "disbursement" | "repayment" | "penalty";
  amount: string;
  created_at: string;
}
